[Loxone] Added null annotations

ppieczul opened the pull request:
* NonNullByDefault was set on all packages except thing factory, which consists of basically single method and results mainly in warning suppression.
  * Also json deserialization classes were marked nullable by default, because they will get nulls wherever a deserialized json does not contain a specific field. However, for the maps in json classes, I added nonnull for both key and value, as I expect GSON to produce a meaningful map or no map at all.
  * This resulted in removal of some dead code and redundant null checks, but also adding missing null checks.
  * To get rid of all compiler warnings, I added suppression only at local variables taking values from legacy calls. Also added some more thread-safe constructions with local object references instead accessing compound data that theoretically could be modified outside of the scope, to eliminate warnings about possible null accesses. They should not happen in reality, but I don't think this is resulting in a bad code.

Signed-off-by: Pawel Pieczul <pieczul@gmail.com> (github: ppieczul)

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific addon: Mention the addon shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the addon README?
- Does your contribution follow the coding guidelines:
  http://docs.openhab.org/developers/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  http://docs.openhab.org/developers/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  http://docs.openhab.org/developers/contributing/contributing#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It's a good practice to add an URL to your built JAR in this Pull Request description,
so it's easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it's reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

ppieczul commented:
@kaikreuzer @martinvw - any chance for a review? Thanks

ppieczul commented:
@kaikreuzer @martinvw - any chance for a review? Thanks

kaikreuzer commented:
@ppieczul We do not have any final agreement on how to use the null annotations - we would like to have some guidelines that are followed in the same way throughout the codebase, so that it is clear to contributors, what is expected from them. We especially didn't (yet) decide to use any of the "*ByDefault". Please find the current discussion [here](https://github.com/eclipse/smarthome/pull/4080#issuecomment-328115241) and feel free to involve yourself in it! (Your PR might be a good reference for pros//cons for the different options).

martinvw commented:
@ppieczul I believe there [a conclusion](https://github.com/eclipse/smarthome/pull/4275/files#diff-24ba7e763168d1631a43ebdef19bcd5d) to the discussion, can you confirm whether this PR already follows those new guidelines.

Thanks!

ppieczul commented:
I will check it over the weekend. Thanks.

On Fri, 22 Sep 2017 at 14:38, Martin van Wingerden <notifications@github.com>
wrote:

> @ppieczul <https://github.com/ppieczul> I believe there a conclusion
> <https://github.com/eclipse/smarthome/pull/4275/files#diff-24ba7e763168d1631a43ebdef19bcd5d>
> to the discussion, can you confirm whether this PR already follows those
> new guidelines.
>
> Thanks!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/2604#issuecomment-331434968>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJx5fqNIj7UGTRSWvo4BFB6oGkFGarksks5sk6ougaJpZM4PD2Oc>
> .
>

ppieczul commented:
@martinvw I think I need your help here. I made a new installation of eclipse IDE and upon running a debug configuration I am getting unresolved upnp bundle probably due to unresolved apache http.

```
!ENTRY org.jupnp 4 0 2017-09-23 23:45:35.981
!MESSAGE FrameworkEvent ERROR
!STACK 0
org.osgi.framework.BundleException: Could not resolve module: org.jupnp [29]
  Unresolved requirement: Import-Package: org.apache.http.client
    -> Export-Package: org.apache.http.client; bundle-symbolic-name="org.apache.httpcomponents.httpclient"; bundle-version="4.5.2.v20170210-0925"; version="4.5.2"; uses:="org.apache.http,org.apache.http.auth,org.apache.http.client.methods,org.apache.http.conn,org.apache.http.conn.routing,org.apache.http.cookie,org.apache.http.params,org.apache.http.protocol"
       org.apache.httpcomponents.httpclient [104]
         Unresolved requirement: Import-Package: org.apache.http; version="[4.4.6,4.5.0)"
  Unresolved requirement: Import-Package: org.apache.http.client.methods
    -> Export-Package: org.apache.http.client.methods; bundle-symbolic-name="org.apache.httpcomponents.httpclient"; bundle-version="4.5.2.v20170210-0925"; version="4.5.2"; uses:="org.apache.http,org.apache.http.client.config,org.apache.http.concurrent,org.apache.http.conn,org.apache.http.message,org.apache.http.params"

	at org.eclipse.osgi.container.Module.start(Module.java:444)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1628)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1608)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1579)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1522)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
```

This prevents from running bindings that use upnp, for example hue which is enabled in the default configuration. I tried purging eclipse and its hidden folders and reinstalling, but same story.

Do you have any ideas how to overcome this?

Thanks
Pawel

ppieczul commented:
First, the problem with unresolved references under IDE seems to be random. It depends on how the eclipse starts. Sometimes after performing all the initial activities, the problem exists (as long as eclipse is open). But sometimes, after subsequent eclipse restart, the problem goes away and everything works fine as long as eclipse session is open. I don't plan to investigate it.

@martinvw I think my changes conform to the general idea of having classes non null by default, with Nullable added wherever it is needed. I added package-info files to each package, as this forces future new classes to follow this guideline.

I updated the PR to the latest code and it should be ready for review/merge (if the checks don't fail...)

maggu2810 commented:
FYI: In ESH we considered not using `package-info` files as it is hard to understand which classes uses nonnull by default and which not then (as you cannot rely on the class annotation only).

ppieczul commented:
If this is the direction, let it be. I change the code to follow it.

ppieczul commented:
@martinvw @kaikreuzer would it be a good time now to review and merge these changes? Thanks.

ppieczul commented:
@martinvw: I merged with the top code. Kindly asking for a review and merge. Thanks.

martinvw commented:
@ppieczul please realise that we have quite some queue and your earlier contributions already jumped the queue, so maybe we can make an arrangement. If you could review some other PR(s) then I will make sure this gets reviewed and merged ASAP. Hereby some (more or less random) suggestions:

* https://github.com/openhab/openhab2-addons/pull/2376
* https://github.com/openhab/openhab2-addons/pull/2350
* https://github.com/openhab/openhab2-addons/pull/2306
* https://github.com/openhab/openhab2-addons/pull/2225
* https://github.com/openhab/openhab2-addons/pull/2348

I always use the following filters to find PR's which are ready to be reviewed: https://github.com/openhab/openhab2-addons/pulls?page=2&q=is%3Apr+is%3Aopen+-label%3A%22Work+in+progress%22+-label%3A%22Awaiting+feedback%22+-assignee%3Akaikreuzer&utf8=%E2%9C%93

martinvw commented on addons/binding/org.openhab.binding.loxone/META-INF/MANIFEST.MF:
```diff
@@ -1,47 +1,47 @@
- org.osgi.framework,
- org.slf4j
-Service-Component: OSGI-INF/*.xml
+Manifest-Version: 1.0
```
This is not a proper resolving of this merge conflict, please follow the new sorting / indentation etc and only add your new imports

martinvw replied:
Did you address this one?

martinvw replied:
Yes you did, only github did not show that properly :-)

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsStateUpdateEvent.java:
```diff
@@ -11,18 +11,22 @@
  * @author Pawel Pieczul - initial contribution
  *
  */
+@NonNullByDefault
```
Would it be less configuration, to not use this one here?

ppieczul replied:
This will result in many incompatibilities in classes that use this class and are non null by default (or we mark all method arguments in this class as non null, which is even more ugly)

martinvw replied:
Okay I did not completely realise the impact 🙂

kaikreuzer commented:
Haha, @ppieczul that's a broad hint that you seem to be a good candidate to help us a bit on reviews. What do you think? It would help making us all faster and reduce the waiting time for everyone - would be cool to have you helping out! 😄

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxUuid.java:
```diff
@@ -32,27 +37,34 @@
+    String cleanupUuid(String uuid) {
+        @SuppressWarnings("null")
+        @NonNull
+        String newId = uuid.replaceAll("[^a-zA-Z0-9-]", "-").toUpperCase();
```
Wouldn't it make more sense to annotate the method and replace the body with:

```
return uuid.replaceAll("[^a-zA-Z0-9-]", "-").toUpperCase();
```

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxServer.java:
```diff
@@ -492,23 +486,36 @@ private void updateConfig(LxJsonApp3 config) {
+                String type = cat.type;
+                if (uuid != null && name != null && type != null) {
+                    addOrUpdateCategory(new LxUuid(uuid), name, type);
+                }
```
See above

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxServer.java:
```diff
@@ -117,28 +123,26 @@ public LxServer(InetAddress host, int port, String user, String password) {
         if (monitorThread == null) {
-            monitorThread = new LxServerThread(this);
-            monitorThread.start();
+            Thread thread = new LxServerThread(this);
```
Found it (again), indeed correct but pretty far fetched, we have a thread running which we cannot stop anymore...

martinvw replied:
Maybe not being able to start it is not such a bad alternative ;-)

ppieczul commented:
@kaikreuzer @martinvw I will be honored to help, as far as my competence and time allows. Let me start with #2350 this weekend.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/handler/LoxoneMiniserverHandler.java:
```diff
@@ -230,12 +256,12 @@ public void onNewConfig(LxServer server) {
-                }
+            channels.addAll(newChannels);
+            for (Channel channel : newChannels) {
+                @SuppressWarnings("null")
```
Channel class seems to be null by default (it has some of the fields annotated with NonNull). But uid can be null and there actually is a (package protected only) default constructor where uid is not initialized. I thing this can be fixed in the Channel class, probably by simply initializing uid to a new ChannelUUID() in the default constructor and annotating this as non null. I will remove this suppression and leave the warning, then try fix it in the framework.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/handler/LoxoneMiniserverHandler.java:
```diff
@@ -201,16 +211,32 @@ public void channelLinked(ChannelUID channelUID) {
+                srv.update(cfg.firstConDelay, cfg.keepAlivePeriod, cfg.connectErrDelay, cfg.responseTimeout,
+                        cfg.userErrorDelay, cfg.comErrorDelay, cfg.maxBinMsgSize, cfg.maxTextMsgSize);
+                srv.start();
+                this.server = srv;
```
Simply using `server` from the class instead of local `srv` gives two warnings that server may be null, which is theoretically right. Using a local reference and assigning it to the class variable once all is initialized looked awkward at first to me. But looking from a more purist point of view, I can incline to saying this is actually the right way to code it so it is thread safe. Not that it has to be thread safe here, but as a general practice. 
Although I must say I'm not sure about your remark that 'it suggests that body construct the server might need to be split of', could you explain it more for me?

martinvw replied:
I don't understand my own sentence to, give me some minutes to re-review that part ;-)

martinvw replied:
What I meant as extracting a method or builder or something which builds the complete server given the configuration parts, this method can then return the complete `LxServer` and it can be assigned straight from that method.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -55,12 +60,16 @@
         this.uuid = uuid;
         if (json.type != null) {
-            this.typeName = json.type.toLowerCase();
+            @SuppressWarnings("null")
```
Makes sense, I will remove warning suppression from all such variables.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -194,11 +203,14 @@ public int hashCode() {
-        this.name = json.name;
+        String jsonName = json.name;
+        if (jsonName != null) {
+            this.name = jsonName;
```
Originally I checked for this null anyway, but later, and substituted it with 'unknown'.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -225,17 +239,20 @@ void update(LxJsonControl json, LxContainer room, LxCategory category) {
                 String value = element.getAsString();
                 if (value != null) {
                     LxUuid id = new LxUuid(value);
+                    @SuppressWarnings("null")
```
It should be better when suppressions are removed from the jdk calls. The ones left will be for GSON and logger. Do you know if there is plan to fix it for the logger? This is common enough to be important.

martinvw replied:
Maybe @triller-telekom knows?

triller-telekom replied:
We are planning to revise our [guidelines](https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style) where we now state that `@NonNullByDefault` should be used per class. depending on the outcome of #4321 we might even disable the warnings in the IDE and only show the error cases.

Regarding third party libraries there is also PR #4217 but that one is waiting for the outcome of #4321, too.

But if you want to keep using the null annotations, please change them to the `@NonNullByDefault` on class level.

Seems like the whole null annotations thing went into business for itself even before we had greed on the above mentioned guidelines. I am sorry if this now hits you so you have to change your annotations.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -225,17 +239,20 @@ void update(LxJsonControl json, LxContainer room, LxCategory category) {
+                        @SuppressWarnings("null")
+                        @NonNull
+                        LxControlState state = states.get(name);
                         state.getUuid().setUpdate(true);
```
Yes, this was an error, or a redundant put changing nothing, the key is already there and we don't change the object reference.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlInfoOnlyAnalog.java:
```diff
@@ -51,12 +54,14 @@
-            format = json.details.format;
-        } else {
-            format = "%.1f";
+        if (json.details != null) {
```
It seems I can revert to the old code without a warning.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxServer.java:
```diff
@@ -349,12 +344,13 @@ public void run() {
             while (running) {
                 // wait until next connect attempt, this time depends on what happened before
-                synchronized (monitorThread) {
+                synchronized (thread) {
```
Makes sense, this should be better monitorThread.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxServer.java:
```diff
@@ -492,23 +486,36 @@ private void updateConfig(LxJsonApp3 config) {
+                String name = room.name;
+                if (uuid != null && name != null) {
+                    addOrUpdateRoom(new LxUuid(uuid), name);
+                }
```
addOrUpdateRoom is a normal flow, I could add a log to the 'else' part, which is absent now.

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/handler/LoxoneMiniserverHandler.java:
```diff
@@ -241,32 +239,14 @@ public void channelLinked(ChannelUID channelUID) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Unknown host");
+        try {
+            server = createNewServer(cfg);
+            if (server == null) {
```
Minor since you are already handling exceptions, maybe you might be better of not returning `null` from `createNewServer()` and throw an exception in that error scenario, then both the OFFLINE markings end up on the same level,

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/LoxoneHandlerFactory.java:
```diff
@@ -24,8 +29,12 @@
+@Component(service = ThingHandlerFactory.class, immediate = true)
 public class LoxoneHandlerFactory extends BaseThingHandlerFactory {
 
+    @SuppressWarnings("null")
```
Maybe you could consider leaving this ones out

Hilbrand replied:
Removing the `@SuppressWarnings` here will give a compiler warning....

ppieczul replied:
`Sets` comes from com.google.common.collect package and we talked about not suppressing only JDK calls, that's why I left it. However, I am ok with removing this and leaving the warning.

martinvw replied:
That is a valid point 👍

triller-telekom replied:
what about dropping the unnecessary dependency to the google library and just use
```
public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>(LoxoneBindingConstants.THING_TYPE_MINISERVER);
```

Hilbrand replied:
Probably better to use: `Collections.singleton()`. Which is also standard java.

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/LoxoneHandlerFactory.java:
```diff
@@ -24,8 +29,12 @@
 public class LoxoneHandlerFactory extends BaseThingHandlerFactory {
 
+    @SuppressWarnings("null")
+    @NonNull
```
Please remove this

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -211,10 +253,11 @@ void update(LxJsonControl json, LxContainer room, LxCategory category) {
 
             for (Map.Entry<String, JsonElement> jsonState : json.states.entrySet()) {
+                @SuppressWarnings("null")
+                @NonNull
```
Please remove all @NonNull

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxJsonApp3.java:
```diff
@@ -21,9 +23,9 @@
-    Map<String, LxJsonControl> controls;
-    Map<String, LxJsonRoom> rooms;
-    Map<String, LxJsonCat> cats;
+    Map<@NonNull String, @NonNull LxJsonControl> controls;
```
Please remove all @NonNull and use @NonNullByDefault

ppieczul replied:
Then I have to add nullable to every field, since they can be null. It also gives me over 20 new warnings due to potential null access, which are these warnings for potential change of the value in multithreaded code, which will never happen here. For example:
```
if (a != null) {
    if (a.b != null) {
        a.b.doSomething();
    }
}
```
This will generate warning at doSomething, because `a` might have changed between `a!=null` and this line.
I think it is not worth it and would rather leave the class not annotated. But then I have to mark the Map key and value and nonnull, because they are never created null, if map is created at all and this seems to break the convention to not use nonnull.
What do you think, which is better? Maybe another solution exists, which I can't see now.

martinvw replied:
I was triggered by the remark of @triller-telekom maybe he knows more.

martinvw replied:
@triller-telekom can you enlighten us?

kaikreuzer replied:
Maybe you want to join in https://github.com/eclipse/smarthome/issues/4321 with these observations/problems. The more we use those annotations, the less we are confident that it is possible to solve all the warnings in a decent way...

martinvw replied:
@kaikreuzer @ppieczul I just read the latest state of the discussion and now I'm in doubt should we just merge this as is. Or do we still want to wait for things to settle down. I think that I read that the usage of @nonnullbydefault is still the prefered way, but not all warnings have to be solved / surpressed?

I think it would be great of one of people actively involved in the discussion could review this PR, maybe @htreu @triller-telekom or @sjka? Thanks all!

ppieczul replied:
It would be great to have this reviewed by more people. 

If there are generally no erroneous changes here, I think (in the spirit of continuous integration) it is more beneficial to merge this and then resolve outstanding warnings at the JDK calls - when they are solved at the framework level. 

I will also summarize my findings in the smarthome disscuscion mentioned above.

With more and more nonnullbydefault added to the framework, I am now getting lots of warnings from this binding too with the version that is on master.
Also, there is a new version of Loxone server software which requires new things to be added to the binding, so it starts stacking up.
Thanks

triller-telekom replied:
I would vote for adding `@NonNullbyDefault` to the class and if your variables can all be `null` you would need to add the `@Nullable` to them.

And please add your findings regarding false positives warnings about the null annotations to our discussion in https://github.com/eclipse/smarthome/issues/4321. The outcome of this discussion will probably be that we change the settings within Eclipse about what we will treat as a warning/error.

For sure we should not block PRs for senseless discussions, but in this case we have the problem that we introduced the null annotations to the framework before we actually agreed on rules on how to use them :( This was a bit unfortunate because while using those annotations they turned out to be rather tricky to use and the tools (like Eclipse) do not support them perfectly. I am sorry that you had to change your PR quite some often because of this, but you were a little to quick for us ;)

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsClient.java:
```diff
@@ -637,24 +653,26 @@ public void onMessage(String msg) {
-                            if (resp != null) {
+                            try {
+                                @SuppressWarnings("null")
+                                @NonNull
```
Please remove all @NonNull and use @NonNullByDefault

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsClient.java:
```diff
@@ -665,11 +683,16 @@ public void onMessage(String msg) {
-                            if (resp != null) {
+                            try {
+                                @SuppressWarnings("null")
+                                @NonNull
```
Please remove all @NonNull and use @NonNullByDefault

martinvw commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsClient.java:
```diff
@@ -686,20 +709,25 @@ public void onMessage(String msg) {
-                            if (config != null) {
+                            try {
+                                @SuppressWarnings("null")
+                                @NonNull
```
Please remove all @NonNull and use @NonNullByDefault

triller-telekom commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/handler/LoxoneMiniserverHandler.java:
```diff
@@ -331,24 +333,53 @@ public void onServerGoesOffline(LxOfflineReason reason, String details) {
         logger.debug("Disposing of server");
-        if (server != null) {
-            server.stop();
+        LxServer srv = server;
```
Why are you copying the class variable into a local one before checking for `null` and stopping the server?

triller-telekom commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -102,11 +139,13 @@ public LxControlState getState(String name) {
     public void dispose() {
-        if (room != null) {
-            room.removeControl(this);
+        LxContainer localRoom = room;
```
Same as above: Why do you copy the class variable into a local one?

triller-telekom commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControl.java:
```diff
@@ -102,11 +139,13 @@ public LxControlState getState(String name) {
         }
-        if (category != null) {
-            category.removeControl(this);
+        LxCategory localCat = category;
```
Same as above: Why do you copy the class variable into a local one?

triller-telekom commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsClient.java:
```diff
@@ -255,22 +260,22 @@ boolean connect() {
             socket = new LxWebSocket();
-            wsClient = new WebSocketClient();
+            WebSocketClient client = new WebSocketClient();
+            wsClient = client;
```
Why do you create a local variable at all?

ppieczul replied:
This is to avoid a warning about potential null access, because theoretically the indirect reference may have changed between the check and use.

‘’’
if (a.b != null) {
    x(a.b)
    y(a.b)
}
‘’’

Apparently in constructions like that compiler will not optimize to cache the referenced value, because this is not what we ask for coding it like that. If the code flow allows for preemption to another thread, the subsequent use of a.b may find b as null. This is my explanation as to why there is a warning. 
This leads to awkward constructions with local variables. I would like to eliminate that and this may lead to warning level tuning but I don’t remember now which one was responsible for that.

triller-telekom replied:
If you just introduced the local variable to satisfy the warnings of the eclipse nullchecker, please undo your changes. We definitely do not want to code around those issue.

But as it is within a `synchronized` block I am assuming that `wsClient` can indeed be changed by another thread, is that correct? If not, please undo your changes as I mentioned above.

ppieczul commented:
I removed the extra variable as @triller-telekom suggested and also made the JSON deserialization classes non null by default.
However, I had to add the variables back in a few cases where mvn reported error (but IDE just warning). I don't really know the difference between these cases and those that only cause warnings. You can see these causing errors in the last commit.

ppieczul commented:
I am putting this PR on hold. I need merging of #2786 first. This will be followed by implementation of new controls and authentication mechanism that were introduced to the most recent version of Loxone Miniserver software. These are more important than null annotations. I will reopen this PR once the other PRs are done.
